standalone/pruntime/enclave/src/libc_hacks.rs

```diff
@@ -0,0 +1,424 @@
+        // man posix_memalign:
+        // posix_memalign() returns zero on success, or one of the error values listed in the next section on failure.  The value of errno is not set.  On Linux (and other systems),
+        // posix_memalign() does not modify memptr on failure.  A requirement standardizing this behavior was added in POSIX.1-2016.
+        if ptr.is_null() && size > 0 {
+            libc::ENOMEM
+        } else {
+            *memptr = ptr;
+            0
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn read(fd: c_int, buf: *mut c_void, count: size_t) -> ssize_t {
+    if fd != 0 {
+        not_allowed!()
+    } else {
+        unsafe { ocall::read(fd, buf, count) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn readv(fd: c_int, iov: *const iovec, iovcnt: c_int) -> ssize_t {
+    assert_eq_size!(libc::iovec, sgx_libc::iovec);
+
+    if fd != 0 {
+        not_allowed!()
+    } else {
+        unsafe { ocall::readv(fd, iov as _, iovcnt) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn write(fd: c_int, buf: *const c_void, count: size_t) -> ssize_t {
+    if !matches!(fd, 1 | 2) {
+        not_allowed!()
+    } else {
+        unsafe { ocall::write(fd, buf, count) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn writev(fd: c_int, iov: *const iovec, iovcnt: c_int) -> ssize_t {
+    assert_eq_size!(libc::iovec, sgx_libc::iovec);
+
+    if !matches!(fd, 1 | 2) {
+        not_allowed!()
+    } else {
+        unsafe { ocall::writev(fd, iov as _, iovcnt) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn getcwd(buf: *mut c_char, size: size_t) -> *mut c_char {
+    // Enclave have no working directory, let's return ""
+    if size > 0 {
+        unsafe { *buf = 0 };
+    }
+    return buf;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn syscall(num: libc::c_long, mut args: ...) -> libc::c_long {
+    if num == libc::SYS_getrandom {
+        let buf = args.arg::<*mut c_void>();
+        let buflen = args.arg::<size_t>();
+        let flags = args.arg::<c_uint>();
+        return getrandom(buf, buflen, flags) as _;
+    }
+    eprintln!("unsupported syscall({})", num);
+    loop {}
+}
+
+#[no_mangle]
+pub extern "C" fn memrchr(s: *mut c_void, c: c_int, n: size_t) -> *mut c_void {
+    unsafe { sgx_libc::memrchr(s as _, c as _, n) as _ }
+}
+
+// sgx_mutex will ignore attributes
+#[no_mangle]
+pub extern "C" fn pthread_mutexattr_init(_attr: *mut libc::pthread_mutexattr_t) -> c_int {
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn pthread_mutexattr_settype(
+    _attr: *mut libc::pthread_mutexattr_t,
+    typ: c_int,
+) -> c_int {
+    // sgx pthread_mutex only support normal mutex, but the stdout() requires PTHREAD_MUTEX_RECURSIVE type.
+    // assert_eq!(typ, libc::PTHREAD_MUTEX_NORMAL);
+
+    static RE: AtomicU16 = AtomicU16::new(0);
+
+    if typ != libc::PTHREAD_MUTEX_NORMAL {
+        // workaround for stdout(), stderr() and stdin()
+        let count = RE.fetch_add(1, Ordering::Relaxed);
+        if count > 3 {
```

Comment:
Also add a `fn libc_hacks_init() { let _ = (stdin(), stdout(), stderr()); }` to simplify the user side codes.